naga/src/back/glsl/mod.rs

```diff
@@ -1186,47 +1190,77 @@ impl<'a, W: Write> Writer<'a, W> {
         Ok(())
     }
 
-    /// Helper method used to write non images/sampler globals
+    /// Helper method used by [Self::write_global] to write just the layout part of
+    /// a non image/sampler global variable, if applicable.
     ///
     /// # Notes
-    /// Adds a newline
     ///
-    /// # Panics
-    /// If the global has type sampler
-    fn write_global(
-        &mut self,
-        handle: Handle<crate::GlobalVariable>,
-        global: &crate::GlobalVariable,
-    ) -> BackendResult {
+    /// Adds trailing whitespace if any layout qualifier is written
+    fn write_global_layout(&mut self, global: &crate::GlobalVariable) -> BackendResult {
+        // Determine which of the `std` layouts we can use (if any)
+        let layout = match global.space {
+            crate::AddressSpace::Uniform => {
+                if self.options.version.supports_std140_layout() {
+                    Some("std140")
+                } else {
+                    None
+                }
+            }
+            crate::AddressSpace::Storage { .. } => {
+                if self.options.version.supports_std430_layout() {
+                    Some("std430")
+                } else if self.options.version.supports_std140_layout() {
+                    Some("std140")
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        };
+
+        // If our version supports explicit layouts, we can also output the explicit binding
+        // if we have it
         if self.options.version.supports_explicit_locations() {
             if let Some(ref br) = global.binding {
                 match self.options.binding_map.get(br) {
                     Some(binding) => {
-                        let layout = match global.space {
-                            crate::AddressSpace::Storage { .. } => {
-                                if self.options.version.supports_std430_layout() {
-                                    "std430, "
-                                } else {
-                                    "std140, "
-                                }
-                            }
-                            crate::AddressSpace::Uniform => "std140, ",
-                            _ => "",
-                        };
-                        write!(self.out, "layout({layout}binding = {binding}) ")?
+                        let layout_padded =
+                            layout.map(|l| format!("{}, ", l)).unwrap_or("".to_string());
+
+                        write!(self.out, "layout({layout_padded}binding = {binding}) ")?;
```

Comment:
This would be nicer if it was 3 `write!` calls. Something like:

```rust
write!(self.out, "layout(")?;
if let Some(layout) = layout {
    write!(self.out, "{}, ", layout)?;
}
write!(self.out, "binding = {binding}) ")?;
```